source/agora/utils/Log.d

```diff
@@ -54,15 +55,88 @@ public template AddLogger (string moduleName = __MODULE__)
 /// Convenience alias
 public alias LogLevel = Level;
 
-version (unittest) {}
-else
+/// Initialize the logger
+static this ()
 {
-    /// Initialize the logger
-    static this ()
-    {
+    version (unittest)
+        auto appender = CircularAppender();
+    else
         auto appender = new AppendConsole();
-        appender.layout(new AgoraLayout());
-        Log.root.add(appender);
+
+    appender.layout(new AgoraLayout());
+    Log.root.add(appender);
+}
+
+/// Circular appender which appends to an internal buffer
+public class CircularAppender : Appender
+{
+    ///
+    private Mask mask_;
+
+    ///
+    private const LineLimit = 100;
+
+    ///
+    private size_t buffer_idx;
+
+    ///
+    private mstring[LineLimit] line_buffer;
```

Comment:
Note: Originally, when I meant circular buffer, I was thinking of just a buffer of 64k. The idea was that GC-allocation can slow things down and hide bugs (see [this](https://preshing.com/20120522/lightweight-in-memory-logging/)).
I don't think it's much of a concern for us at the moment, or ever will be, just wanted to add perspective.

Reply:
Right. I was worried the output would be cut mid-sentence and might cause misleading error logs. But 64k should be big enough.

Allow me to rework the PR? :)